subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java

```diff
@@ -52,19 +96,22 @@ static ServiceRegistry createWorkerDaemonServices(ServiceRegistry parent) {
     @Override
     public DefaultWorkResult execute(ActionExecutionSpec spec) {
         try {
-            Worker worker = getIsolatedClassloaderWorker(spec.getClassLoaderStructure());
+            ServiceRegistry workServices = new WorkerPublicServicesBuilder(new RelativeFileWorkerServices(spec.getBaseDir(), internalServices))
+                    .withInternalServicesVisible(spec.isInternalServicesRequired())
+                    .build();
+            Worker worker = getIsolatedClassloaderWorker(spec.getClassLoaderStructure(), workServices);
```

Comment:
I'm not quite sure, but is there a worker reuse problem here?

Let's say that we have two projects that are running the same kind of work actions.  First task starts, submits some work and that spins up a worker.   Second task starts, submits some work and another worker is spun up.  There's lots of work left and it can bounce around between the two workers.

The two workers are identical classpath wise, but the baseDir is different between them.  Previously, all of the File-like things were shipped over as absolute paths, but now with these services, something like...

```
fileSystemOperations.copy {
   from("here")
   into("there")
}
```
means something different depending on the value of this baseDir.   I think we're trying to make this baseDir == project dir, but the current working directory for workers is somewhere under GRADLE_USER_HOME.

Does that make sense?  This looks like something worth testing for.

Reply:
Yep, you're right on.  We're caching the worker because everything it required used to be the same between work items (otherwise you'd get a different worker daemon), but now the services the worker gets could be different, so we need to handle the caching differently.  Good catch.

Reply:
This should be fixed now (with a test to cover it).